drivers/gles3/storage/material_storage.cpp

```diff
@@ -39,32 +39,24 @@
 
 #include "drivers/gles3/rasterizer_canvas_gles3.h"
 #include "drivers/gles3/rasterizer_gles3.h"
+#include "servers/rendering/storage/variant_converters.h"
 
 using namespace GLES3;
 
 ///////////////////////////////////////////////////////////////////////////
 // UBI helper functions
 
-_FORCE_INLINE_ static void _fill_std140_variant_ubo_value(ShaderLanguage::DataType type, int p_array_size, const Variant &value, uint8_t *data) {
+static void _fill_std140_variant_ubo_value(ShaderLanguage::DataType type, int p_array_size, const Variant &value, uint8_t *data) {
 	switch (type) {
 		case ShaderLanguage::TYPE_BOOL: {
 			uint32_t *gui = (uint32_t *)data;
 
 			if (p_array_size > 0) {
-				const PackedInt32Array &ba = value;
-				int s = ba.size();
-				const int *r = ba.ptr();
-
-				for (int i = 0, j = 0; i < p_array_size; i++, j += 4) {
-					if (i < s) {
-						gui[j] = (r[i] != 0) ? 1 : 0;
-					} else {
-						gui[j] = 0;
-					}
-					gui[j + 1] = 0; // ignored
-					gui[j + 2] = 0; // ignored
-					gui[j + 3] = 0; // ignored
+				PackedInt32Array ba = value;
+				for (int i = 0; i < ba.size(); i++) {
+					ba.set(i, ba[i] ? 1 : 0);
```

Comment:
Is this necessary? The code seems highly suspicious. It looks like it is copying over the entire ``value`` array, then iterating over it to clean the values, then passing the whole array to a new function to copy each value one-by-one into another array. Could we not convert, clean, and copy all at once as we do now?

Its nice that this templated approach cuts the amount of code in half, but I have a feeling it is forcing us to do way more copies and way more operations to do the same amount of work.

Reply:
Note that old code does convert+copy+iterate passes too, they are just harder to see. In the old way, these lines:

```cpp
const PackedInt32Array &ba = value;
```

create a new temporary: a PackedInt32Array from a value variant object using Variant conversion. It just binds it as a const reference, but a temporary is still made and allocated. So a full conversion and array allocation is done here, anyway, it's just harder to see. The following manual for-loop is replaced by `write_array_std140(`), but they do the same thing. The amount of work done should be roughly the same between the old and new code. And when passed the same values, the new code does not really do any more conversions than the old one, there is a fast path in `convert_array_std140()` for many packed arrays. The only extra copying comes when passing more complex object arrays which have to allocate some temporary extra space, but these didn't work at all in the old code. I suspect because doing these conversions for arbitrary arrays and objects properly is actually pretty hard, as evident. I though about special casing all types and use internal variant accessors in variant_internal.h to handle these conversions without temporary copies, but it turned out that kind of code is even harder to write, read and understand.

The only exception here is indeed the boolean uniform handling (bool, bvec2, bvec3, bvec4), which has to do extra work for the "force to 1 or 0"-loop. However, the only extra cost is the loop (which is typically much less than 100 elements), because the temporary COW packed array is already made and its CowData reference count is 1, which means it doesn't have to copy anything when written into because it is already a non-shared copy which had to be made anyway (both in old and new versions).

In practice, it shouldn't have much effect on performance, but uniform boolean conversions are probably is a bit slower than the original version for the same arguments. Of course, I doubt many people actually use uniform boolean in the first place. However, to match the original performance and not make copy of the COW vector, I could write:

```cpp
const PackedInt32Array &ba = convert_array_std140<Vector2i, int32_t>(value);
write_array_std140<Vector2i>(ba, gui, p_array_size, 4);
for (int = 0; i < p_array_size * 4) {
    gui[i] = gui[i] ? 1 : 0;
}
```

It still makes one extra loop over the data, but as uniform arrays are not usually very large in practice it shouldn't matter too much. Alternatively, we could add an optional lambda function converter to inline the conversion into the copy loop, which the compiler can optimize nicely:

```cpp
const PackedInt32Array &ba = convert_array_std140<Vector2i, int32_t>(value);
write_array_std140<Vector2i>(ba, gui, p_array_size, 4, [](int32_t value) { return value ? 1: 0 });
```

With this version, the performance should be pretty much identical. Or we could leave the new code as-is, it's pretty simple to understand this way with a pretty minimal performance penalty.

Reply:
Thank you for the detailed response. I didn't realize that ``const PackedInt32Array &ba = value;`` was copying the entire array.

Indeed the boolean arrays are the worst offenders in terms of copying necessarily, but they are also the cheapest to copy and likely won't be that large. 

I'm happy with the current state of things in that case. I would avoid making it any more complicated. Already this PR adds significant complexity by adding layers of templated functions. I understand that was done in order to make handling all the edge cases manageable, but adding more complexity will make this impossible to debug six months from now.